### PR TITLE
#3545 Raise an exception when applying operations on shared tensors that requires crypto_provider

### LINKED ISSUE #3545: Raise an exception when crypto_provider is not provided when sharing the tensor
**Is your feature request related to a problem? Please describe.**
When we share the tensor using share() method, it by default assigns the calling worker (me) as a crypto provider. This allows us to perform addition and subtraction, however, it raises ObjectNotFound error after multiplication.   

**Describe the solution you'd like**
Raise warning when share() is called, something like 
>  You haven't provided crypto_provider. You won't be able to perform certain operations like multiplication

And a specific error when multiply or equate something like `CryptoProviderNotProvided` 

**To replicate**

```python

x = torch.tensor([1,2,3]).share(alice, bob)
y = torch.tensor([1,2,3]).share(alice, bob)

```

```python
print((x+y).get()) # Works
print((x*y).get()) # Raises ObjectNotFoundError

```

### PR DESCRIPTION
**Pull request by ariannmichael**
## Description
Created a new exception for when multiply or compare shared tensor without crypto_provider
and a warning when a tensor is shared without crypto_provider

Thank you for your contribution to the PySyft repository.
Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context.

If your pull request closes a GitHub issue, then set its number below.

Resolves #3545 (issue)

## Checklist:
* [ ] My changes are covered by tests.
* [x] I have run [the pre-commit hooks](https://github.com/OpenMined/PySyft/blob/master/CONTRIBUTING.md#setting-up-pre-commit-hook) to format and check my code for style issues.
* [x] I have commented my code following [Google style](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html).

(See the [the contribution guidelines](https://github.com/OpenMined/PySyft/blob/master/CONTRIBUTING.md) for additional tips.)

### CODE REVIEW

**Comment by LaRiffle** [comment]
Hey! I had a thought about this: this should not raise an exception as long as the parties don't need to do mltiplcation or operations which require a crypto provider. 
So I would suggest: set the crypto_provide to None is it's not provided and raise an error if people try to do multiplications.

The idea behind is that for secure aggregatin we're just doing a mean so we don't need to have a crypto provider

**Comment by ariannmichael** [comment]
I'm sorry my title was misleading, the exception is only raised when operations that requires the crypto_provider are called, like multiplication.

File: syft/frameworks/torch/tensors/interpreters/additive_shared.py [hunk 1]
```diff
@@ -116,6 +117,11 @@ def __init__(
         # max value for shares in field
         self._max_value = None
 
+        if crypto_provider is None:
+            warnings.warn(
+                "You haven't provided crypto_provider. You won't be able to perform certain operations like multiplication"
+            )
+
         self.crypto_provider = (
             crypto_provider if crypto_provider is not None else sy.hook.local_worker
```
Comment (LaRiffle) [inline syft/frameworks/torch/tensors/interpreters/additive_shared.py:126]: You can remove this:
```
self.crypto_provider = (
            crypto_provider if crypto_provider is not None else sy.hook.local_worker	            crypto_provider if crypto_provider is not None else sy.hook.local_worker
```
to keep crypto_provider  = None otherwise your next check won't work

File: syft/exceptions.py [hunk 2]
```diff
@@ -369,6 +369,12 @@ def __init__(self, crypto_store, crypto_type, available_instances, n_instances):
         super().__init__(message)
 
 
+class CryptoProviderNotFoundError(Exception):
+    """Raised when trying to multiply or equate shared tensors without crypto provider."""
+
+    pass
```
Comment (LaRiffle) [inline syft/exceptions.py:375]: I ket you add an explanation message :)

**Comment by LaRiffle** [comment]
Can you make sure it fails with an explicit message for common operations?
This would solve things like: https://github.com/OpenMined/PySyft/issues/3540
You can use decorator like `@crypto_provider_required` if it helps to factorize code (not 100% sure)

File: syft/frameworks/torch/mpc/__init__.py [hunk 3]
```diff
@@ -29,6 +31,7 @@ def decorator(f):
         name = f.__qualname__
         protocol_store[(name, protocol_name)] = f
 
+        @wraps(f)
```
Comment (youben11) [inline syft/frameworks/torch/mpc/__init__.py:34]: Why is this required?

Comment (ariannmichael) [reply syft/frameworks/torch/mpc/__init__.py:34]: For using 2 decorators on same function.

**Comment by ariannmichael** [comment]
Yes, I'll do it

**Comment by gmuraru** [comment]
Hey @ariannmichael. Are you still working on this?

**Comment by gmuraru** [comment]
Ping @ariannmichael.
